### DISCUSSION TITLE
Audio and video options not getting enabled

### DISCUSSION THREAD

**Discussion by amoljagdalepucsd** [Ylianst/MeshCentral#5797] [Issues]
Hi Community ,

I am currently running Meshcentral [1.0.85] in hybrid mode and have activated the chat feature on a remote machine, which is within a different private network in my home. The chat functionality is fully operational, including audio and video options. However, when a friend located approximately 1000 km away in a different city attempts to access the same machine for a session, remote desktop session working fine but these options are not enabled in chat. I am unsure whether this issue stems from a network or firewall problem, or if I am missing some port configurations in the config.JSON file. Or do i need to change mode insted of hybrid  in config.JSON, please suggest me. Additionally, I checked on the Mesh community channel and found the link (https://github.com/Ylianst/MeshCentral/issues/4388) about STUN servers and TURN servers. How can I ensure that this is the same issue affecting me?

Config.JSON
{
	"$schema": "
[https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json"](https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json%22)
,
	"settings": {
		"cert": "abc.com",
		"_WANonly": false,
		"_LANonly": false,
		"_sessionKey": "MyReallySecretPassword1",
		"sessionSameSite": "lax",
		"port": 4430,
		"aliasPort": 443,
		"_redirPort": 800,
		"_redirAliasPort": 80,
		"agentPort": 2121,
                "agentAliasPort": 443,
		"relayPort": 453,
		"_relayAliasPort": 453,
		"agentPortTls": false,
                "agentPing": 10,
		"browserPing": 10,
		"agentIdleTimeout": 18000,
		"tlsOffload": true,
		"SelfUpdate": false,
		"allowFraming": true,
		"allowLoginToken": true,
		"webRTC": true,
		"relayDNS": true,
                "multitenant":false,
		"amtscanner": false,
		"amtManager": false,
		"plugins": {
			"enabled": true
		},
		"postgres": {
			"user": "xyz",
			"password": "12345",
			"port": "5432",
			"host": "123.1.2.5",
			"database": "meshdb",
			"ssl": false
		},
		"dbExpire": {
			"events": 0,
			"powerevents": 0,
			"statsevents": 0
		},
		"agentCoreDump": true,
		"agentLogDump": true,
		"minify": true,
		"clipboardGet": true,
		"clipboardSet": true,
		"compression": true,
		"wsCompression": true,
		"agentWsCompression": true,
		"noAgentUpdate": 1,
		"temporaryAgentUpdate": false,
		"agentsInRam": true,
		"desktopMultiplex": true
	},
	"domains": {
		"": {
			"title": "MeshCentral",
			"title2": "Mesh service",
			"minify": "true",
			"NewAccounts": false,
			"_userNameIsEmail": true,
			"certUrl": "abc.com",
			"tlsOffload": "true",
			"notifyimage": "mesh.png",
			"meshMessengerTitle": "meshmessenger",
			"nightmode": 1,
			"userSessionIdleTimeout": 90    
		}
	}
}

**Comment by si458** [comment]
the code says this
```
QV('camButton', webchannel && webchannel.ok && !localStream && (userMediaSupport == 2));
QV('micButton', webchannel && webchannel.ok && !localStream && (userMediaSupport > 0));
```
so the buttons will ONLY show
1. IF you have enabled webrtc
2. webrtc was OK and connected
3. IF you arent already using using the webcam/microphone elsewhere
4. IF the user have given the web browser permission to use the microphone/webcam and the browser supports it

so best to check ur friend, what browser he is using, and if he denied any permissions that popped up asking to use webcam/microphone

**Reply by amoljagdalepucsd** [reply]
Yes @si458 I already checked this and webrtc/webchannel is not connecting and its value is null.
we enabled browser permissions and userMediaSupport value [ userMediaSupport == 2 ]to check compatibility is also correct.

**Reply by si458** [reply]
in theory it will use google and mozilla as default stun servers,
HOWEVER i you might be having problems trying to connect to them so you might need to add your own to help create the webrtc connection
try adding the following to your `settings` section in `config.json`
```
"webrtcConfig": {
	"iceServers": [
	  {"urls": "stun:STUNSERVERHERE.COM"},
	  {"urls": "stun:2NDSTUNSERVERHERE.COM"}
	]
}
 ```
OR
as @Ylianst explains here https://github.com/Ylianst/MeshCentral/issues/4388#issuecomment-1214180487 try using [openrelay TURN servers](https://www.metered.ca/tools/openrelay/), but u will need an account now (still free tho)

**Reply by amoljagdalepucsd** [reply]
Ok @si458 , 
I am checking it in more details. still do you know any developer who can help in this?

**Reply by si458** [reply]
whats the output of `webchannel` from the debug window of the `https://yourserver.com/messenger` ?
mine looks like this
`RTCDataChannel { ok: true, label: "DataChannel", negotiated: false, ordered: true, reliable: true, maxPacketLifeTime: null, maxRetransmits: null, protocol: "", id: 1, readyState: "open" }`
also i spotted that every so often the mic/cam button wouldnt show BUT a refresh of the webpage, then it appears

**Reply by amoljagdalepucsd** [reply]
For me, once the WebRTC/webchannel connection happens properly and options are enabled, there's no need to refresh; they always appear. 

Let me share the WebChannel details with you when the options are not getting enabled.

**Reply by amoljagdalepucsd** [reply]
Hi @si458 
I have conducted a more detailed investigation into the issue. I attempted to resolve it by using my own STUN server, but unfortunately, it did not yield any results. Additionally, I tried using Google's/Mozilla servers and metered STUN/TURN servers with my account, as suggested by @Ylianst in (https://github.com/Ylianst/MeshCentral/issues/4388.) However, I encountered the same issue.

Subsequently, I discovered that these options are only enabled when both systems are connected via mobile networks. If one system is connected via WiFi or LAN, the options remain disabled. Please review the configuration settings or advise if I am overlooking something

**Reply by amoljagdalepucsd** [reply]
Hi @si458 , Do you have any thoughts on this ?
**Subsequently, I discovered that these audio and video options are only enabled when both systems are connected via mobile networks. If one system is connected via WiFi or LAN, the options remain disabled. Please review the configuration settings or advise if I am overlooking something.**

**Reply by amoljagdalepucsd** [reply]
Hi @si458 , I am facing below issue with latest Meshcentral as well. how to resolve this ?
**Subsequently, I discovered that these audio and video options are only enabled when both systems are connected via mobile networks. If one system is connected via WiFi or LAN, the options remain disabled. Please review the configuration settings or advise if I am overlooking something.**
